### PR TITLE
KAFKA-3587: Fix fake large log segment in log cleaner

### PR DESCRIPTION
**Pull request by Ishiihara**

### CODE REVIEW

File: core/src/main/scala/kafka/log/OffsetMap.scala [hunk 1]
```diff
@@ -30,6 +30,8 @@ trait OffsetMap {
   def clear()
   def size: Int
   def utilization: Double = size.toDouble / slots
+  def notFull: Boolean
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/OffsetMap.scala:33]: Generally, positive naming is preferable (i.e. `isFull`).

**Comment by alekar** [comment]
The code in this PR means that offset returned by `buildOffsetMapForSegment` can terminate at a point which is not on segment boundaries. How does that work with `clean` which seems to assume that `buildOffsetMap` terminated at a segment boundary?

**Comment by Ishiihara** [comment]
@alekar We will try to fit the first segment in the map. For subsequent segments, we use offset difference as an upper bound of the number of messages and will not add the messages in the segment to the map. This ensures that we always stop at a segment boundary or throw an exception if the map cannot fit the first segment.

**Comment by alekar** [comment]
@Ishiihara isn't that difference in segment offsets the bad heuristic causing this issue? Also, it has no way of accounting for duplicate ratio.

**Comment by Ishiihara** [comment]
@alekar Agreed. I think the fix is to make the minimum change to account for fixing the issue. Also, if we ends up not adding the segment to the map in this round, we will add the segment in the next round. 

I agree that reading the segment can make a better estimate, but it also have additional IO cost and memory overhead.  In terms of performance, it really depends on the key distribution in a segment.

**Comment by alekar** [comment]
@Ishiihara right. If you look at what I had in mind (on JIRA), it trades off how many segments it can compact in one run. I'm happy to take this up as an improvement if a quick fix is required at this moment.

However, I am skeptical that moving the use of offset difference out to the second segment seen by `buildOffsetMap` is a true fix.

**Comment by edoardocomar** [comment]
@Ishiihara - the very pessimistic check
"val segmentSize = segment.nextOffset() - segment.baseOffset"
is still present in this PR

**Comment by Ishiihara** [comment]
@alekar @edoardocomar The fix will solve the issue in the JIRA as we alway make sure that the first segment fit in the map with the its actual size.  But I aggre it does not solve the map utilization problem. As I said, this is a minimum change to address the JIRA. I will close this PR and waiting for @alekar's PR .

**Comment by guozhangwang** [comment]
@alekar Is there already a PR or could you describe your solution a bit more here?

**Comment by alekar** [comment]
https://github.com/apache/kafka/pull/1329 I will fix Integration test failures after we decide if that `take all or no offsets from segment` approach is necessary on JIRA.

**Pull request closed without merging**